api/v1beta1/oscmachine_validation.go

```diff
@@ -0,0 +1,243 @@
+package v1beta1
+
+import (
+	"errors"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"regexp"
+	"strings"
+)
+
+const (
+	minIops = 0
+	maxIops = 13000
+	minSize = 0
+	maxSize = 14901
+)
+
+// ValidateOscMachineSpec validate each parameters of OscMachine spec
+func ValidateOscMachineSpec(spec OscMachineSpec) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.Node.Vm.KeypairName != "" {
+		if errs := ValidateVmKeypairName(spec.Node.Vm.KeypairName, field.NewPath("keypairName")); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+	if spec.Node.Vm.ImageId != "" {
+		if errs := ValidateVmImageId(spec.Node.Vm.ImageId, field.NewPath("imageId")); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+	if spec.Node.Vm.DeviceName != "" {
+		if errs := ValidateVmDeviceName(spec.Node.Vm.DeviceName, field.NewPath("deviceName")); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+	if spec.Node.Vm.VmType != "" {
+		if errs := ValidateVmVmType(spec.Node.Vm.VmType, field.NewPath("vmType")); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+	if spec.Node.Vm.SubregionName != "" {
+		if errs := ValidateNetworkSubregionName(spec.Node.Vm.SubregionName, field.NewPath("subregionName")); len(errs) > 0 {
+			allErrs = append(allErrs, errs...)
+		}
+	}
+
+	if len(spec.Node.Volumes) != 0 {
+		volumesSpec := spec.Node.Volumes
+		for _, volumeSpec := range volumesSpec {
+			if volumeSpec.Iops != 0 {
+				if errs := ValidateVolumeIops(volumeSpec.Iops, field.NewPath("iops")); len(errs) > 0 {
+					allErrs = append(allErrs, errs...)
+				}
+			}
+
+			if volumeSpec.Size != 0 {
+				if errs := ValidateVolumeSize(volumeSpec.Size, field.NewPath("size")); len(errs) > 0 {
+					allErrs = append(allErrs, errs...)
+				}
+			}
+
+			if volumeSpec.VolumeType != "" {
+				if errs := ValidateVolumeVolumeType(volumeSpec.VolumeType, field.NewPath("volumeType")); len(errs) > 0 {
```

Comment:
why for types, you write ressources name twice ?

Reply:
i change it.
To distinguish ValidateVolumeVolumeType (ValidateNodeVolumeType now )to ValidateVolumeType